### PR TITLE
Correctly  handle padding in JS2

### LINKED ISSUE #656: _intoFFI padding
Should double check with LLVM output, but basically:

```rs
struct Foo {
  a : u8,
  b : u8,
  c : u32
}
```

Pads with `2 x i8`s. But

```rs
struct Bar {
  b : u16,
  c : u32
}
```

Pads with `1 x i16`. The padding is determined by the previous type, and then it all gets converted into `i32`s for WASM, regardless of how much sense that makes.

So, the fix should go in `generate_fields`, to grab what the previous field's align is. Then update the padding to be use something like `(next_offset - curr_offset - field_layout.size()) / previous_align`

### LINKED ISSUE #657: Rust Wasm ABI treats two-field structs like scalars and doesn't require padding
When single-field structs are passed down over Wasm, they're treated as a single parameter. Multi-field structs are passed down as multiple parameters.

Except not quite: multi-field structs can have _padding_. A struct `(u8, u32)` has 3 bytes of padding after the `u8`, represented in llvm as `[3 x i8]`. `(u16, u32)` has two bytes of padding, represented in llvm as `[1 x i16]`.

In the JS/Wasm backend, of course we need to deal with padding when deciphering _returned_ structs (since we read them from memory with their layout). However, we also need to deal with padding when passing structs over Wasm, because each entry of padding is one argument. 

This means that in the following setup:

```rust
#[repr(C)]
struct Foo {
  a: u8,
  // three bytes of padding
  b: u32,
  c: u32,
}

extern "C" fn foo(x: Foo) {}
``

`foo()` gets called as `foo(a, 0, 0, 0, b, c)`.

We already have this behavior implemented in https://github.com/rust-diplomat/diplomat/pull/643, except currently it assumes all padding bytes are `i8`, whereas it appears that they match the lowest alignment of the previous unaligned section. Or something like that (`(u8, u8, u32)` and `(u16, u32)` have differently segmented padding sections). This is a separate bug to be fixed (https://github.com/rust-diplomat/diplomat/issues/656).


**However**, there appears to be strangeness even when dealing with padding. The [Wasm tool conventions](https://github.com/WebAssembly/tool-conventions/blob/main/BasicCABI.md) claim that structs with single scalar fields are passed as scalars, and structs with multiple are passed as aggregates, however in the following code:

```rust
#[repr(C)]
pub struct Scalar {
    first: u32,
}

#[repr(C)]
pub struct Small {
  a: u8,
  // 3 bytes padding
  b: u32
}

#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub struct Big {
    a: u8,
    b: bool,
    c: u8,
    // 1 byte padding
    d: u32,
    e: i32,
}

#[no_mangle]
pub extern "C" fn scalar(x: Scalar) {}
#[no_mangle]
pub extern "C" fn small(x: Small) {}
#[no_mangle]
pub extern "C" fn big(x: Big) {}
```

the resultant LLVM IR looks something like this:

```rust
%Scalar = type { i32 }
%Big = type { i8, i8, i8, [1 x i8], i32, i32 }

define dso_local void @scalar(%Scalar %0) unnamed_addr #0 {...}

define dso_local void @small(i8 %x.0, i32 %x.1) unnamed_addr #0 {...}

define dso_local void @big(%Big %0) unnamed_addr #0 {...}
```

where `Small` is split into two scalar parameters in the function. The function `small()` in WASM is correctly invoked as `small(a, b)`, with no padding.

With a bunch of experimentation it seems like this consistently happens for structs that (transitively) contain two scalar fields. The sizes of the fields are irrelevant, but adding a third field always turns it into a separate `type {...}` thing that may contain padding.

We should figure out where this heuristic is coming from, and either get it fixed (it doesn't match the documented conventions) or write code to handle it. This may unfortunately require making `_intoFFI()` conditional on cases with and without padding, but that won't be too hard.

### PR DESCRIPTION
**Pull request by Manishearth**
This PR:

 - Tracks the *type* of padding field, generating the right number of padding fields when the type is wider than i8 (fixes #656)
 - Handles the edge case that scalar fields are passed as separate arguments without padding *unless found in a larger struct*. (fixes #657)

It also generally cleans up the padding code and makes it reusable.

### CODE REVIEW

**Comment by Manishearth** [comment]
Forgot to handle end-padding .Done now.

File: feature_tests/js/api/ScalarPairWithPadding.mjs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    #fromFFI(ptr) {
+        const firstDeref = (new Uint8Array(wasm.memory.buffer, ptr, 1))[0];
+        this.#first = firstDeref;
+        const secondDeref = (new Uint32Array(wasm.memory.buffer, ptr + 4, 1))[0];
+        this.#second = secondDeref;
+    }
+
+    assertValue() {
+        let functionCleanupArena = new diplomatRuntime.CleanupArena();
+        wasm.ScalarPairWithPadding_assert_value(...this._intoFFI());
```
Comment (sffc) [inline feature_tests/js/api/ScalarPairWithPadding.mjs:63]: Question: the WASM ABI expects a padding zero for every byte, not just for every i32?

File: tool/src/js/type_generation/mod.rs [hunk 2]
```diff
@@ -217,35 +215,57 @@ impl<'ctx, 'tcx> TyGenContext<'ctx, 'tcx> {
...
+                _ if !matches!(&field.ty, &hir::Type::Struct(_)) => ForcePaddingStatus::NoForce,
+                // 2-field struct contained in 2-field struct, caller decides
+                (2, 2) => {
+                    needs_force_padding = true;
+                    ForcePaddingStatus::PassThrough
+                }
+                // Outer struct has > 3 fields, always pad
+                (2, 3..) => ForcePaddingStatus::Force,
+                // Larger fields will always have padding anyway
+                _ => ForcePaddingStatus::NoForce
```
Comment (sffc) [inline tool/src/js/type_generation/mod.rs:234]: Thought: Wow, very specific rules. I trust this is the result of your research

File: tool/src/js/layout.rs [hunk 3]
```diff
@@ -57,24 +117,32 @@ pub fn unit_size_alignment() -> Layout {
...
+) -> (Layout, usize) {
     match typ {
         // repr(C) fieldless enums use the default platform representation: isize
-        Type::Enum(..) => Layout::new::<usize_target>(),
-        Type::Opaque(..) => opaque_size_alignment(),
-        Type::Slice(..) => Layout::new::<(usize_target, usize_target)>(),
-        Type::Primitive(p) => primitive_size_alignment(*p),
+        Type::Enum(..) => (Layout::new::<usize_target>(), 1),
+        Type::Opaque(..) => (opaque_size_alignment(), 1),
+        Type::Slice(..) => (Layout::new::<(usize_target, usize_target)>(), 2),
```
Comment (sffc) [inline tool/src/js/layout.rs:132]: Thought: Okay, "scalar" doesn't mean "a primitive value", it means "number of u32s". ?

File: tool/src/js/layout.rs [hunk 4]
```diff
@@ -11,38 +11,98 @@ use diplomat_core::hir::{
...
 
+pub struct StructFieldLayout {
+    /// The offset of this field in the struct
+    pub offset: usize,
+    /// The number of padding fields needed after this field
+    pub padding_count: usize,
+    /// The size of the padding field
+    pub padding_size: usize,
+    /// The number of scalar fields in this field
+    pub scalar_count: usize,
```
Comment (sffc) [inline tool/src/js/layout.rs:22]: Nit: Maybe define what you mean when you say "scalar"

**Review by sffc — APPROVED** [review]

File: feature_tests/js/api/ScalarPairWithPadding.mjs (hunk 1 above)
Comment (Manishearth) [reply feature_tests/js/api/ScalarPairWithPadding.mjs:63]: The padding fields can have different sizes, ranging from i8 to i32, based on the preceding field. That's what the first half of this PR fixes.

File: tool/src/js/type_generation/mod.rs (hunk 2 above)
Comment (Manishearth) [reply tool/src/js/type_generation/mod.rs:234]: Yeah

File: tool/src/js/layout.rs (hunk 3 above)
Comment (Manishearth) [reply tool/src/js/layout.rs:132]: No, it's integral primitives. Slices aren't primitives.

**Pull request merged**